pkg/activator/net/revision_backends_test.go

```diff
@@ -1522,67 +1522,296 @@ func TestServiceMoreThanOne(t *testing.T) {
 	}
 
-	// Initial state for tests
-	cur := dests{
-		ready:    sets.NewString("10.10.1.3", "10.10.1.2"),
-		notReady: sets.NewString("10.10.1.1"),
+	type test struct {
+		name     string
+		input    input
+		expected expected
 	}
 
-	// Test all pods healthy (skips probes)
-	rw.healthyPods = cur.ready.Union(cur.notReady)
-	healthy, noop, notMesh, err := rw.probePodIPs(cur.ready, cur.notReady)
-	if !healthy.Equal(rw.healthyPods) {
-		t.Errorf("Healthy does not match, got %#v, want %#v diff: %s", healthy, rw.healthyPods, cmp.Diff(healthy, cur.ready))
-	}
-	if !noop || notMesh || err != nil {
-		t.Errorf("Unexpected values: noop=%t (true), notMesh=%t (false), err=%s (nil)", noop, notMesh, err)
+	tests := []test{
+		{
+			name: "all healthy", // Test skipping probes when all endpoints are healthy
+			input: input{
+				current: dests{
+					ready:    sets.NewString("10.10.1.1"),
+					notReady: sets.NewString("10.10.1.2"),
+				},
+				healthy: sets.NewString("10.10.1.1", "10.10.1.2"),
+			},
+			expected: expected{
+				healthy:   sets.NewString("10.10.1.1", "10.10.1.2"),
+				noop:      true,
+				notMesh:   false,
+				success:   true,
+				numProbes: 0,
+			},
+		},
+		{
+			name: "one pod fails probe", // Test that we probe all pods when one fails
+			input: input{
+				current: dests{
+					notReady: sets.NewString("10.10.1.1", "10.10.1.2", "10.10.1.3"),
+				},
+				hostResponses: map[string][]activatortest.FakeResponse{
+					"10.10.1.1": {{
+						Err: errors.New("podIP transport error"),
+					}},
+					"10.10.1.2": {{
+						Code:  http.StatusOK,
+						Body:  queue.Name,
+						Delay: 250 * time.Millisecond,
+					}},
+				},
+				enableProbeOptimization: true,
+			},
+			expected: expected{
+				healthy:   sets.NewString("10.10.1.2", "10.10.1.3"),
+				noop:      false,
+				notMesh:   true,
+				success:   false,
+				numProbes: 3,
+			},
+		},
+		{
+			name: "ready pods skipped without mesh auto",
+			input: input{
+				current: dests{
+					ready:    sets.NewString("10.10.1.1"),
+					notReady: sets.NewString("10.10.1.2"),
+				},
+				enableProbeOptimization: true,
+				meshMode:                netcfg.MeshCompatibilityModeDisabled,
+			},
+			expected: expected{
+				healthy:   sets.NewString("10.10.1.1", "10.10.1.2"),
+				noop:      false,
+				notMesh:   true,
+				success:   true,
+				numProbes: 1,
+			},
+		},
+		{
+			name: "ready pods not skipped with mesh auto",
+			input: input{
+				current: dests{
+					ready:    sets.NewString("10.10.1.1"),
+					notReady: sets.NewString("10.10.1.2"),
+				},
+				enableProbeOptimization: true,
+				meshMode:                netcfg.MeshCompatibilityModeAuto,
+			},
+			expected: expected{
+				healthy:   sets.NewString("10.10.1.1", "10.10.1.2"),
+				noop:      false,
+				notMesh:   true,
+				success:   true,
+				numProbes: 2,
+			},
+		},
+		{
+			name: "only ready pods healthy without probe optimization", // NOTE: prior test is effectively this one with probe optimization
```

Comment:
**non blocking question:** 

I guess to confirm - probe optimization=false doesn't have an effect and probing happens

I thought probe optimization when disabled would not probe at all - (since the revision has a container with an exec probe and the queue proxy can't intercept the container's healthy state)

Reply:
The probe optimization in this `probePodIPs` context just allows non-ready pods to be marked as healthy after successful probe.  Without the optimization enabled the pod needs to have a successful probe and be ready to be marked healthy.  Interestingly enough it still probes the pod regardless.  Definitely could update to just only probe the ones it could actually mark healthy... but I don't really want to start adding changes unrelated to the fix here.

I think part of the current behavior has to do possibly with the mesh mode Auto behavior for Mesh detection.  Since it needs to probe to detect.